### ISSUE TITLE
Primary flow re-establishes as a subflow but immediately closes

### ISSUE THREAD

**Issue by arinc9** [multipath-tcp/mptcp_net-next#512]
Both hosts run the export/20240726T103611 tree with v5 of "mptcp: fix endpoints with 'signal' and 'subflow' flags" patch series applied.

```
root@OpenWrt:/# ip mp e
192.168.11.250 id 1 subflow dev eth1 
41.147.192.176 id 2 subflow dev wwan1 
100.116.95.86 id 3 subflow dev wwan0 

[         CREATED] token=36c011e6 remid=0 locid=0 saddr4=192.168.11.250 daddr4=102.132.169.219 sport=54746 dport=1194
[14368.596516] MPTCP: check if should create a new subflow
[14368.596516] msk->pm.local_addr_used: 0, local_addr_max: 3
[14368.596516] msk->pm.subflows: 0, subflows_max: 5
[14368.613137] MPTCP: saddr 41.147.192.176: __mptcp_subflow_connect() ran
[14368.621150] MPTCP: saddr 100.116.95.86: __mptcp_subflow_connect() ran
[     ESTABLISHED] token=36c011e6 remid=0 locid=0 saddr4=192.168.11.250 daddr4=102.132.169.219 sport=54746 dport=1194
[  SF_ESTABLISHED] token=36c011e6 remid=0 locid=2 saddr4=41.147.192.176 daddr4=102.132.169.219 sport=57835 dport=1194 backup=0 ifindex=10
[  SF_ESTABLISHED] token=36c011e6 remid=0 locid=3 saddr4=100.116.95.86 daddr4=102.132.169.219 sport=36189 dport=1194 backup=0 ifindex=9

root@OpenWrt:/# ip mp e d i 1
[       SF_CLOSED] token=36c011e6 remid=0 locid=0 saddr4=192.168.11.250 daddr4=102.132.169.219 sport=54746 dport=1194 backup=0

root@OpenWrt:/# ip mp e a 192.168.11.250 subflow dev eth1
[14387.061458] MPTCP: check if should create a new subflow
[14387.061458] msk->pm.local_addr_used: 1, local_addr_max: 3
[14387.061458] msk->pm.subflows: 2, subflows_max: 5
[14387.077556] MPTCP: saddr 192.168.11.250: __mptcp_subflow_connect() ran
[  SF_ESTABLISHED] token=36c011e6 remid=0 locid=4 saddr4=192.168.11.250 daddr4=102.132.169.219 sport=38375 dport=1194 backup=0 ifindex=3
[       SF_CLOSED] token=36c011e6 remid=0 locid=4 saddr4=192.168.11.250 daddr4=102.132.169.219 sport=38375 dport=1194 backup=0 ifindex=3
```

Other paths won't be closed after being re-established. This is not an issue with the eth1 path; if it's not the primary flow, it won't be closed after being re-established.

**Comment by matttbe** [comment]
Hi @arinc9,

Thank you for the bug report

What are the limits on both sides? (It looks like it is 5 on the client side, but what about the server side?)

It is a bit confusing, but the limits are for the "additional subflows". It means that if you have 2 additional subflows, and you remove and re-add the endpoint used by the initial subflow, the new created subflow will be counting as an additional one. So in this case, if the subflow limits are set to 2, that's fine, until you recreate the endpoint used by the initial subflow: there you need the limit to be at least 3.

**Comment by matttbe** [comment]
Also, do you have the same issue if you force re-using the same ID? (what should be done in theory, to re-use the 'local ID 0'):

```
ip mptcp endpoint add 192.168.11.250 subflow dev eth1 id 1
```

**Comment by pabeni** [comment]
@arinc9: please also include `nstat -as *Tcp*` output for both client and server

**Comment by matttbe** [comment]
@arinc9: I found another issue when removing the ID0 where the `RM_ADDR` might not be sent as expected for the ID0 (and there might be another bug where connections could be half closed when the RM_ADDR is not sent before, I'm still looking at that). Do you mind trying this patch please? (quickly made, not tested yet)

[rm_id_0.patch.txt](https://github.com/user-attachments/files/16418097/rm_id_0.patch.txt)

**Comment by matttbe** [comment]
@arinc9: there were some issues in my previous patch. This should be applied on top: 
[rm_id_0_part2.patch.txt](https://github.com/user-attachments/files/16426237/rm_id_0_part2.patch.txt)

I'm still investigating stuff, proper patches will come later on

**Comment by arinc9** [comment]
@matttbe these two diffs resolved the issue, thank you so much! Let me know if you want further testing.

```
# ip mp e
192.168.11.254 id 1 subflow dev eth1 
192.168.8.162 id 3 subflow dev wan 
7.0.229.159 id 4 subflow dev wwan0 
100.88.43.153 id 5 subflow dev wwan1 
# ip mp e d i 1
[ 6279.487769] MPTCP: matttbe: removing addr id 0 (addr: 1) -> ret:0 force:1
[       SF_CLOSED] token=df0e889d remid=0 locid=0 saddr4=192.168.11.254 daddr4=102.132.169.219 sport=42074 dport=1194 backup=0
[       SF_CLOSED] token=df0e889d remid=0 locid=0 saddr4=192.168.11.254 daddr4=102.132.169.219 sport=42074 dport=1194 backup=0
# ip mp e a 192.168.11.254 subflow dev eth1
[  SF_ESTABLISHED] token=df0e889d remid=0 locid=6 saddr4=192.168.11.254 daddr4=102.132.169.219 sport=57155 dport=1194 backup=0 ifindex=3
```

**Comment by matttbe** [comment]
Great, thank you for having tested! I need to cleanup that and send proper patches. Once it is done, it would be great if you could validate the new series :)

**Comment by arinc9** [comment]
Sure, I'll leave my tested-by. CC me!

**Comment by matttbe** [comment]
@arinc9 sorry, I forgot to CC you, but the new series has been sent on our ML: https://lore.kernel.org/all/20240802-mptcp-pm-avail-v6-0-964ba9ce279f@kernel.org/T/#u

Feel free to send your Tested-By tag(s)

**Comment by arinc9** [comment]
@matttbe sorry, I'm a bit late. Both hosts run export/20240809T121429 with [this series](https://patchwork.kernel.org/project/mptcp/list/?series=878199) applied. When adding the endpoint manually a second time, a subflow won't be established using the endpoint. Same behaviour if I add the subflow without an ID, remove it, then add it without an ID again. We did not have this behaviour with your custom patch.

```
[         CREATED] token=14952930 remid=0 locid=0 saddr4=192.168.11.254 daddr4=102.132.169.219 sport=53304 dport=1194
[     ESTABLISHED] token=14952930 remid=0 locid=0 saddr4=192.168.11.254 daddr4=102.132.169.219 sport=53304 dport=1194
[  SF_ESTABLISHED] token=14952930 remid=0 locid=3 saddr4=41.147.196.209 daddr4=102.132.169.219 sport=37441 dport=1194 backup=0 ifindex=10
[  SF_ESTABLISHED] token=14952930 remid=0 locid=4 saddr4=100.123.85.199 daddr4=102.132.169.219 sport=48285 dport=1194 backup=0 ifindex=11

# ip mp e
192.168.11.254 id 1 subflow dev eth1 
41.147.196.209 id 3 subflow dev wwan0 
100.123.85.199 id 4 subflow dev wwan1 

# ip mp e d i 1
[       SF_CLOSED] token=14952930 remid=0 locid=0 saddr4=192.168.11.254 daddr4=102.132.169.219 sport=53304 dport=1194 backup=0
[       SF_CLOSED] token=14952930 remid=0 locid=0 saddr4=192.168.11.254 daddr4=102.132.169.219 sport=53304 dport=1194 backup=0

# ip mp e a 192.168.11.254 id 1 subflow dev eth1 
[  SF_ESTABLISHED] token=14952930 remid=0 locid=0 saddr4=192.168.11.254 daddr4=102.132.169.219 sport=43253 dport=1194 backup=0 ifindex=3

# ip mp e d i 1
[       SF_CLOSED] token=14952930 remid=0 locid=0 saddr4=0.0.0.0 daddr4=102.132.169.219 sport=53304 dport=0 backup=0
[       SF_CLOSED] token=14952930 remid=0 locid=0 saddr4=192.168.11.254 daddr4=102.132.169.219 sport=43253 dport=1194 backup=0 ifindex=3

# ip mp e a 192.168.11.254 id 1 subflow dev eth1 
#
```

**Comment by matttbe** [comment]
@arinc9 : thank you for having tested!

Indeed, the 're-re-creation' of the ID 0 endpoint was not working. I think it was a regression from my previous version.

Please note that in the 'ip mptcp monitor' output, we can see that the SF_CLOSED event is wrongly generated twice. Sometimes with no source address and destination port.

All of that should be fixed with this new series: https://lore.kernel.org/mptcp/20240815-mptcp-dup-close-evt-v1-0-5a551d3a66cc@kernel.org/T/

Do you mind checking that please?

**Comment by arinc9** [comment]
@matttbe thank you. Re-re-creation works now. However, deleting the endpoint for the third time results in a warning. Other than that, a subflow using the endpoint will be created every time so my issue is resolved. Let me know when should I close this issue.

```
[         CREATED] token=8912598c remid=0 locid=0 saddr4=192.168.11.254 daddr4=102.132.169.219 sport=42318 dport=1194
[     ESTABLISHED] token=8912598c remid=0 locid=0 saddr4=192.168.11.254 daddr4=102.132.169.219 sport=42318 dport=1194
[  SF_ESTABLISHED] token=8912598c remid=0 locid=2 saddr4=41.147.196.10 daddr4=102.132.169.219 sport=33549 dport=1194 backup=0 ifindex=10
[  SF_ESTABLISHED] token=8912598c remid=0 locid=5 saddr4=100.114.191.98 daddr4=102.132.169.219 sport=47439 dport=1194 backup=0 ifindex=11

root@OpenWrt:~# ip mp e
192.168.11.254 id 1 subflow dev eth1 
41.147.196.10 id 2 subflow dev wwan1 
100.114.191.98 id 5 subflow dev wwan2 

root@OpenWrt:~# ip mp e d i 1
[       SF_CLOSED] token=8912598c remid=0 locid=0 saddr4=192.168.11.254 daddr4=102.132.169.219 sport=42318 dport=1194 backup=0

root@OpenWrt:~# ip mp e a 192.168.11.254 subflow dev eth1
[  SF_ESTABLISHED] token=8912598c remid=0 locid=0 saddr4=192.168.11.254 daddr4=102.132.169.219 sport=34047 dport=1194 backup=0 ifindex=3

root@OpenWrt:~# ip mp e
41.147.196.10 id 2 subflow dev wwan1 
100.114.191.98 id 5 subflow dev wwan2 
192.168.11.254 id 6 subflow dev eth1 

root@OpenWrt:~# ip mp e d i 6
[       SF_CLOSED] token=8912598c remid=0 locid=0 saddr4=192.168.11.254 daddr4=102.132.169.219 sport=34047 dport=1194 backup=0 ifindex=3

root@OpenWrt:~# ip mp e a 192.168.11.254 subflow dev eth1
[  SF_ESTABLISHED] token=8912598c remid=0 locid=0 saddr4=192.168.11.254 daddr4=102.132.169.219 sport=32827 dport=1194 backup=0 ifindex=3

root@OpenWrt:~# ip mp e
41.147.196.10 id 2 subflow dev wwan1 
100.114.191.98 id 5 subflow dev wwan2 
192.168.11.254 id 6 subflow dev eth1 

root@OpenWrt:~# ip mp e d i 6
[  467.461639] ------------[ cut here ]------------
[  467.466263] WARNING: CPU: 2 PID: 9243 at __mark_subflow_endp_available+0x64/0x6c
[  467.473658] Modules linked in:
[       SF_CLOSE[  467.476706] CPU: 2 UID: 0 PID: 9243 Comm: ip Not tainted 6.11.0-rc2 #0
D] token=8912598[  467.484602] Hardware name: Bananapi BPI-R4 (DT)
c remid=0 locid=[  467.490509] pstate: 40400005 (nZcv daif +PAN -UAO -TCO -DIT -SSBS BTYPE=--)
0 saddr4=192.168[  467.498840] pc : __mark_subflow_endp_available+0x64/0x6c
.11.254 daddr4=1[  467.505522] lr : mptcp_nl_remove_subflow_and_signal_addr.isra.0+0x2dc/0x330
02.132.169.219 s[  467.513853] sp : ffffffc08b86b7b0
port=32827 dport[  467.518541] x29: ffffffc08b86b7f0 x28: ffffff80c3fdb300 x27: ffffff80c4804600
=1194 backup=0 i[  467.527045] x26: ffffffc080919818 x25: ffffffc080dafa40 x24: 0000000000000000
findex=3
[  467.535549] x23: ffffff80c1151580 x22: 0000000000000001 x21: ffffff80c1151590
[  467.543533] x20: ffffff80c3fdb360 x19: ffffff80c3fdad00 x18: 0000000000000000
[  467.550657] x17: 000a000600003b80 x16: 00090006dba98466 x15: 00070008fe0ba8c0
[  467.557782] x14: ffffffffffffffff x13: 0000000000000020 x12: 0101010101010101
[  467.564906] x11: 7f7f7f7f7f7f7f7f x10: fefefefefefefeff x9 : ffffff80c0abc000
[  467.572032] x8 : 0400000000000000 x7 : 0000000000000000 x6 : ffffff80ff790ae8
[  467.579157] x5 : ffffffffffffffdb x4 : 0000000000000000 x3 : ffffff80c3fdb400
[  467.586282] x2 : ffffffffffffff9b x1 : 0000000000000000 x0 : ffffff80c3fdad00
[  467.593407] Call trace:
[  467.595843]  __mark_subflow_endp_available+0x64/0x6c
[  467.600798]  mptcp_pm_nl_del_addr_doit+0x164/0x218
[  467.605578]  genl_family_rcv_msg_doit+0xac/0x10c
[  467.610186]  genl_rcv_msg+0x1b0/0x244
[  467.613839]  netlink_rcv_skb+0x5c/0x128
[  467.617666]  genl_rcv+0x34/0x48
[  467.620797]  netlink_unicast+0x1e4/0x2d8
[  467.624709]  netlink_sendmsg+0x16c/0x36c
[  467.628620]  ____sys_sendmsg+0x1c8/0x278
[  467.632536]  ___sys_sendmsg+0x7c/0xc0
[  467.636188]  __sys_sendmsg+0x44/0x98
[  467.639752]  __arm64_sys_sendmsg+0x20/0x28
[  467.643838]  invoke_syscall.constprop.0+0x4c/0xd8
[  467.648533]  do_el0_svc+0x3c/0xb8
[  467.651837]  el0_svc+0x18/0x54
[  467.654882]  el0t_64_sync_handler+0x118/0x124
[  467.659228]  el0t_64_sync+0x150/0x154
[  467.662880] ---[ end trace 0000000000000000 ]---
```

**Comment by matttbe** [comment]
> Re-re-creation works now. However, deleting the endpoint for the third time results in a warning. 

Thanks! I managed to reproduce it as well. Should be fixed in the v2.

https://lore.kernel.org/mptcp/20240816-mptcp-dup-close-evt-v2-0-8a33f6617f5c@kernel.org/T/

**Comment by arinc9** [comment]
Perfect. I don't encounter any issues whatsoever with establishing flows anymore. Thank you very much! You can close the issue if it's time.

**Comment by matttbe** [comment]
Thank you for having tested! (sorry, I forgot to report your Tested-by).

> You can close the issue if it's time.

It will be automatically closed when the patches will be applied.
